server/handlers/validation.go

```diff
@@ -192,71 +201,78 @@ func loadTargetsFromStore(gun, role string, repo *tuf.Repo, store storage.MetaSt
+	var prev *data.SignedSnapshot
+	_, currentJSON, err := store.GetCurrent(gun, data.CanonicalSnapshotRole)
+	if err == nil {
+		prev = new(data.SignedSnapshot)
+		if err = json.Unmarshal(currentJSON, prev); err != nil {
+			logrus.Error("Failed to unmarshal existing snapshot for GUN ", gun)
+			return nil, err
+		}
 	}
 
-	algo, keyBytes, err := store.GetKey(gun, data.CanonicalSnapshotRole)
-	if err != nil {
-		return nil, validation.ErrBadHierarchy{Msg: "could not retrieve snapshot key. client must provide snapshot"}
+	if _, ok := err.(storage.ErrNotFound); !ok && err != nil {
+		return nil, err
 	}
-	foundK := data.NewPublicKey(algo, keyBytes)
 
-	validKey := false
-	for _, id := range role.ListKeyIDs() {
-		if id == foundK.ID() {
-			validKey = true
-			break
-		}
-	}
-	if !validKey {
+	metaUpdate, err := snapshot.NewSnapshotUpdate(prev, repo)
+	switch err.(type) {
+	case signed.ErrInsufficientSignatures, signed.ErrNoKeys:
+		// If we cannot sign the snapshot, then we don't have keys for the snapshot,
+		// and the client should have submitted a snapshot
 		return nil, validation.ErrBadHierarchy{
 			Missing: data.CanonicalSnapshotRole,
 			Msg:     "no snapshot was included in update and server does not hold current snapshot key for repository"}
-	}
+	case nil:
+		return metaUpdate, nil
 
-	_, currentJSON, err := store.GetCurrent(gun, data.CanonicalSnapshotRole)
-	if err != nil {
-		if _, ok := err.(storage.ErrNotFound); !ok {
-			return nil, validation.ErrValidation{Msg: err.Error()}
-		}
+	default:
+		return nil, validation.ErrValidation{Msg: err.Error()}
 	}
-	var sn *data.SignedSnapshot
-	if currentJSON != nil {
-		sn = new(data.SignedSnapshot)
-		err := json.Unmarshal(currentJSON, sn)
-		if err != nil {
-			return nil, validation.ErrValidation{Msg: err.Error()}
-		}
-		err = repo.SetSnapshot(sn)
-		if err != nil {
-			return nil, validation.ErrValidation{Msg: err.Error()}
-		}
-	} else {
-		// this will only occur if no snapshot has ever been created for the repository
-		err := repo.InitSnapshot()
-		if err != nil {
-			return nil, validation.ErrBadSnapshot{Msg: err.Error()}
+}
+
+// generateTimestamp generates a new timestamp from the previous one in the store - this assumes all
+// the other roles have already been set on the repo, and will set the generated timestamp on the repo as well
+func generateTimestamp(gun string, repo *tuf.Repo, store storage.MetaStore) (*storage.MetaUpdate, error) {
+	var prev *data.SignedTimestamp
+	_, currentJSON, err := store.GetCurrent(gun, data.CanonicalTimestampRole)
+	if err == nil {
+		prev = new(data.SignedTimestamp)
+		if err = json.Unmarshal(currentJSON, prev); err != nil {
+			logrus.Error("Failed to unmarshal existing timestamp for GUN ", gun)
+			return nil, err
 		}
 	}
-	sgnd, err := repo.SignSnapshot(data.DefaultExpires(data.CanonicalSnapshotRole))
-	if err != nil {
-		return nil, validation.ErrBadSnapshot{Msg: err.Error()}
+
+	if _, ok := err.(storage.ErrNotFound); !ok && err != nil {
```

Comment:
I can definitely move this above the `if err == nil`, but that `if` cannot be de-indented because this handles the case where the error is not `storage.ErrNotFound`, and that `if` handles the case where there is no error at all, but we basically want to do nothing if the error was `storage.ErrNotFound`.

I can change this to a switch statement though, since it's somewhat confusing.